nixos/modules/security/wrappers/wrapper.c

```diff
@@ -151,9 +152,55 @@ static int make_caps_ambient(const char *self_path) {
     return 0;
 }
 
+// These are environment variable aliases for glibc tunables.
+// This list shouldn't grow further, since this is a legacy mechanism.
+// Any future tunables are expected to only be accessible through GLIBC_TUNABLES.
+//
+// They are not included in the glibc-provided UNSECURE_ENVVARS list,
+// since any SUID executable ignores them. This wrapper also serves
+// executables that are merely granted ambient capabilities, rather than
+// being SUID, and hence don't run in secure mode. We'd like them to
+// defend those in depth as well, so we clear these explicitly.
+//
+// Except for MALLOC_CHECK_ (which is marked SXID_ERASE), these are all
+// marked SXID_IGNORE (ignored in secure mode), so even the glibc version
+// of this wrapper would leave them intact.
+#define UNSECURE_ENVVARS_TUNABLES \
+    "MALLOC_CHECK_\0" \
+    "MALLOC_TOP_PAD_\0" \
+    "MALLOC_PERTURB_\0" \
+    "MALLOC_MMAP_THRESHOLD_\0" \
+    "MALLOC_TRIM_THRESHOLD_\0" \
+    "MALLOC_MMAP_MAX_\0" \
+    "MALLOC_ARENA_MAX\0" \
+    "MALLOC_ARENA_TEST\0"
+
 int main(int argc, char **argv) {
     ASSERT(argc >= 1);
 
+    int debug = getenv(wrapper_debug) != NULL;
+
+    // Drop insecure environment variables explicitly
+    //
+    // glibc does this automatically in SUID binaries, but we'd like to cover this:
+    //
+    //  a) before it gets to glibc
+    //  b) in binaries that are only granted ambient capabilities by the wrapper,
+    //     but don't run with an altered effective UID/GID, nor directly gain
+    //     capabilities themselves, and thus don't run in secure mode.
+    //
+    // We're using musl, which doesn't drop environment variables in secure mode,
```

Comment:
The kernel code that sets AT_SECURE sets it in three cases: UID/GID ≠ effective UID/GID, file capability bits (but not ambient caps, as discussed), and LSMs requesting it.

I don't have much experience with LSMs, but I think it would be very worthwhile to research either figuring out how to configure an existing LSM to cover this, or writing a small LSM of our own.

Having a single entry point for legitimate privilege escalation is a security boon in many ways, but the ambient capability case throws a bit of a wrench in the works. If we cover it with an LSM, we should be golden.

That said, I'd like to make sure we don't block _this_ PR on that work; it's a pre-existing issue that is neither worsened nor improved by this change.

Reply:
If we actually wanted an LSM for this, I think it would be much better to make one that would allow for simulation of file capabilities.

Reply:
I think some new syscall allowing for the capabilities to be passed on, and setting `AT_SECURE` if any permitted capability is ambient, would be the best kernel-level solution. But this is really somewhat orthogonal to this PR, and it's probably only a fix in 10 years anyway.